crates/turbo-tasks-memory/src/scope.rs

```diff
@@ -194,115 +200,92 @@ impl TaskScope {
         let old_count = self.unfinished_tasks.fetch_sub(1, Ordering::AcqRel);
         debug_assert!(old_count > 0);
         if old_count == 1 {
-            let value = backend.flag_scope_change();
-            let _ = self.last_task_finish_generation.fetch_update(
-                Ordering::Relaxed,
-                Ordering::Relaxed,
-                |v| {
-                    if v < value {
-                        Some(value)
-                    } else {
-                        None
-                    }
-                },
-            );
-            self.event.notify(usize::MAX);
+            self.update_unfinished_state(backend);
         }
     }
 
-    pub fn has_unfinished_tasks(
-        &self,
-        self_id: TaskScopeId,
-        backend: &MemoryBackend,
-    ) -> Option<EventListener> {
-        'restart: loop {
-            // There would be a race condition when we check scopes in a certain
-            // order. e. g. we check A before B, without locking both at the same
-            // time. But it can happen that a change propagates from B to A in the
-            // meantime, which means we would miss the unfinished work. In this case
-            // we would not get the strongly consistent guarantee. To counter that
-            // we introduce a global generation counter, which is incremented before
-            // checking. When a task finishes (resp. unfinished_tasks is decreased) we must
-            // also update the local generation counter to the global one. When
-            // all tasks are finished the scope can no longer influence the unfinished work
-            // of other scopes. By ensuring that all work has been finished before the start
-            // of checking the whole tree, we ensure that all scopes are either done, or
-            // contain unfinished work.
-            // Note that a change can propagate into any direction: from parent to child,
-            // from child to parent and from siblings. Also through multiple
-            // layers.
-            let start_generation = backend.acquire_scope_generation();
-            let mut checked_scopes = HashSet::new();
-            if self.unfinished_tasks.load(Ordering::Acquire) != 0 {
-                let listener = self.event.listen();
-                if self.unfinished_tasks.load(Ordering::Relaxed) != 0 {
-                    return Some(listener);
+    fn decrement_unfinished_tasks_internal(&self) -> bool {
+        self.unfinished_tasks.fetch_sub(1, Ordering::AcqRel) == 1
+    }
+
+    pub fn add_parent(&self, parent: TaskScopeId, backend: &MemoryBackend) {
+        let mut state = self.state.lock();
+        if state.parents.add(parent) && state.has_unfinished_tasks {
+            backend.with_scope(parent, |parent| {
+                let update = parent.increment_unfinished_tasks_internal();
+
+                // lock need to include increment_unfinished_tasks
+                drop(state);
```

Comment:
Does this create a race condition? Perhaps we should create an internal method which expects an already held lock?